test/model/typing_status_test.dart

```diff
@@ -0,0 +1,223 @@
+      checkNotifiedOnce();
+
+      handleTypingEvent(dmNarrow, TypingOp.stop, eg.otherUser);
+      checkTypists({groupNarrow: [eg.otherUser]});
+      checkNotNotified();
+    });
+
+    test('remove one of two typists in the same narrow', () {
+      prepareModel(typistsByNarrow: {
+        groupNarrow: [eg.otherUser, eg.thirdUser],
+      });
+
+      handleTypingEvent(groupNarrow, TypingOp.stop, eg.otherUser);
+      checkTypists({groupNarrow: [eg.thirdUser]});
+      checkNotifiedOnce();
+    });
+
+    test('remove all typists in a narrow', () {
+      prepareModel(typistsByNarrow: {dmNarrow: [eg.otherUser]});
+
+      handleTypingEvent(dmNarrow, TypingOp.stop, eg.otherUser);
+      checkTypists({});
+      checkNotifiedOnce();
+    });
+
+    test('remove typists from different narrows', () {
+      prepareModel(typistsByNarrow: {
+        dmNarrow: [eg.otherUser],
+        groupNarrow: [eg.thirdUser],
+        topicNarrow: [eg.fourthUser],
+      });
+
+      handleTypingEvent(groupNarrow, TypingOp.stop, eg.thirdUser);
+      checkTypists({dmNarrow: [eg.otherUser], topicNarrow: [eg.fourthUser]});
+      checkNotifiedOnce();
+
+      handleTypingEvent(dmNarrow, TypingOp.stop, eg.otherUser);
+      checkTypists({topicNarrow: [eg.fourthUser]});
+      checkNotifiedOnce();
+
+      handleTypingEvent(topicNarrow, TypingOp.stop, eg.fourthUser);
+      checkTypists({});
+      checkNotifiedOnce();
+    });
+  });
+
+  group('cancelling old timer', () {
+    test('when typing stopped early', () => awaitFakeAsync((async) async {
+      prepareModel();
+
+      handleTypingEvent(dmNarrow, TypingOp.start, eg.otherUser);
+      checkTypists({dmNarrow: [eg.otherUser]});
+      checkNotifiedOnce();
+      check(async.pendingTimers).length.equals(1);
+
+      handleTypingEvent(dmNarrow, TypingOp.stop, eg.otherUser);
+      checkTypists({});
+      checkNotifiedOnce();
+      check(async.pendingTimers).isEmpty();
+    }));
+
+    test('when typing repeatedly started', () => awaitFakeAsync((async) async {
+      prepareModel();
+
+      handleTypingEvent(groupNarrow, TypingOp.start, eg.otherUser);
+      checkTypists({groupNarrow: [eg.otherUser]});
+      checkNotifiedOnce();
+      check(async.pendingTimers).length.equals(1);
+
+      // The new timer should be active and the old timer should be cancelled.
+      handleTypingEvent(groupNarrow, TypingOp.start, eg.otherUser);
+      checkTypists({groupNarrow: [eg.otherUser]});
+      check(async.pendingTimers).length.equals(1);
+      checkNotNotified();
+    }));
+  });
+
+  group('typing start expiry period', () {
+    test('typist is removed when the expiry period ends', () => awaitFakeAsync((async) async {
+      prepareModel();
+
+      handleTypingEvent(dmNarrow, TypingOp.start, eg.otherUser);
+      async.elapse(const Duration(seconds: 5));
+      checkTypists({dmNarrow: [eg.otherUser]});
+      async.elapse(const Duration(seconds: 10));
+      checkTypists({});
+    }));
```

Comment:
should also `checkNotifiedOnce` here

and for good measure, `checkNotNotified` at the 5-second mark

Reply:
Also at the end check pendingTimers is empty. (No need for checking in the middle that it's nonempty ­— the fact that `async.elapse` causes a spontaneous change is enough to make that point.)

Reply:
A related update: for the revised "repeated typing start event resets the timer"  test, I kept some `pendingTimers` checks. It probably makes more sense there because we don't want duplicated timers.